main/src/library/Map.flix

```diff
@@ -82,35 +54,24 @@ namespace Map {
     ///
     /// Returns `true` if and only if `m` contains the key `k`.
     ///
-    pub def memberOf(k: k, m: Map[k, v]): Bool = match m {
-        case Map(Nil) => false
-        case Map((a, _) :: rs) => if (a == k) true else memberOf(k, Map(rs))
-    }
+    pub def memberOf(k: k, m: Map[k, v]): Bool =
+        let Map(xs) = m;
+        RBTree.contains(k, xs)
 
     ///
     /// Returns the keys of `m`.
     ///
-    pub def keysOf(m: Map[k, v]): Set[k] =
-        let Map(xs) = m;
-            List.toSet(fst(List.unzip(xs)))
+    pub def keysOf(m: Map[k, v]): Set[k] = foldLeftWithKey((k, b, _) -> Set.insert(k, b), Set.empty(), m)
 
     ///
     /// Returns the values of `m`.
     ///
-    pub def valuesOf(m: Map[k, v]): List[v] =
-        let Map(xs) = m;
-            snd(List.unzip(xs))
-
-
-
-
-
+    pub def valuesOf(m: Map[k, v]): List[v] = foldLeftWithKey((k, b, _) -> k :: b, Nil, m)
```

Comment:
Does this return in the order want? Im not sure. I would have expected a foldRight.

Reply:
foldRight was indeed the correct choice here.